landoapi/hgexports.py

```diff
@@ -393,6 +393,147 @@ def get_timestamp(self) -> str:
+        r"bug [0-9]+",
+        r"no bug",
+        r"^(back(ing|ed)?\s+out|backout).*(\s+|\:)[0-9a-f]{12}",
+        r"^(revert(ed|ing)?).*(\s+|\:)[0-9a-f]{12}",
+        r"^add(ed|ing)? tag",
+    ]
+]
+INVALID_REVIEW_FLAG_RE = re.compile(r"[\s.;]r\?(?:\w|$)")
+
+CHANGESET_KEYWORD = r"(?:\b(?:changeset|revision|change|cset|of)\b)"
+CHANGESETS_KEYWORD = r"(?:\b(?:changesets|revisions|changes|csets|of)\b)"
+SHORT_NODE = r"([0-9a-f]{12}\b)"
+SHORT_NODE_RE = re.compile(SHORT_NODE, re.I)
+BACKOUT_KEYWORD = r"^(?:backed out|backout|back out)\b"
+BACKOUT_KEYWORD_RE = re.compile(BACKOUT_KEYWORD, re.I)
+BACKOUT_SINGLE_RE = re.compile(
+    BACKOUT_KEYWORD
+    + r"\s+"
+    + CHANGESET_KEYWORD
+    + r"?\s*"
+    + r"(?P<node>"
+    + SHORT_NODE
+    + r")",
+    re.I,
+)
+BACKOUT_MULTI_SPLIT_RE = re.compile(
+    BACKOUT_KEYWORD + r"\s+(?P<count>\d+)\s+" + CHANGESETS_KEYWORD, re.I
+)
+BACKOUT_MULTI_ONELINE_RE = re.compile(
+    BACKOUT_KEYWORD
+    + r"\s+"
+    + CHANGESETS_KEYWORD
+    + r"?\s*"
+    + r"(?P<nodes>(?:(?:\s+|and|,)+"
+    + SHORT_NODE
+    + r")+)",
+    re.I,
+)
+RE_SOURCE_REPO = re.compile(r"^Source-Repo: (https?:\/\/.*)$", re.MULTILINE)
+RE_SOURCE_REVISION = re.compile(r"^Source-Revision: (.*)$", re.MULTILINE)
+# Like BUG_RE except it doesn't flag sequences of numbers, only positive
+# "bug" syntax like "bug X" or "b=".
+BUG_CONSERVATIVE_RE = re.compile(r"""(\b(?:bug|b=)\b(?:\s*)(\d+)(?=\b))""", re.I | re.X)
+BUG_RE = re.compile(
+    r"""# bug followed by any sequence of numbers, or
+        # a standalone sequence of numbers
+         (
+           (?:
+             bug |
+             b= |
+             # a sequence of 5+ numbers preceded by whitespace
+             (?=\b\#?\d{5,}) |
+             # numbers at the very beginning
+             ^(?=\d)
+           )
+           (?:\s*\#?)(\d+)(?=\b)
+         )""",
+    re.I | re.X,
+)
+
+
+def is_backout(commit_desc: str) -> bool:
+    """Returns True if commit description indicates the changeset is a backout.
+
+    Backout commits should always result in is_backout() returning True,
+    and parse_backouts() not returning None.  Malformed backouts may return
+    True here and None from parse_backouts().
+    """
+    return BACKOUT_KEYWORD_RE.match(commit_desc) is not None
+
+
+def parse_bugs(commit_desc: str, conservative: bool = False) -> list[int]:
+    m = RE_SOURCE_REPO.search(commit_desc)
+    if m:
+        source_repo = m.group(1)
+
+        if source_repo.startswith("https://github.com/"):
+            conservative = True
+
+    if commit_desc.startswith("Bumping gaia.json"):
+        conservative = True
+
+    bugzilla_re = BUG_CONSERVATIVE_RE if conservative else BUG_RE
+
+    bugs_with_duplicates = [int(m[1]) for m in bugzilla_re.findall(commit_desc)]
+    bugs_seen = set()
+    bugs_seen_add = bugs_seen.add
+    bugs = [x for x in bugs_with_duplicates if not (x in bugs_seen or bugs_seen_add(x))]
+    return [bug for bug in bugs if bug < 100000000]
```

Comment:
Might be a good opportunity to add a comment describing what 100000000 is referring to here.